### PR TITLE
feat: add imperative streaming API with tail-window optimization

### PR DESCRIPTION
**Pull request by jaysin586**
## Summary

Add an imperative streaming API (`writeChunk`/`resetStream`) to SvelteMarkdown, giving developers direct control over chunk-by-chunk rendering for LLM streaming UIs. This avoids prop reactivity edge cases with identical consecutive chunks and enables websocket-style offset patching. Also ships a tail-window reparsing optimization that skips re-lexing stable prefix tokens, significantly reducing parse cost for long streams.

## Changes

✨ **New features**
- Imperative `writeChunk()` and `resetStream()` exported methods on `SvelteMarkdown`
- Append mode (string chunks) and offset mode (`{ value, offset }` patches)
- Mode locking per stream session with clear warnings on misuse
- `StreamingChunk` and `StreamingOffsetChunk` types exported from `index.ts`
- Batched append flushing via `requestAnimationFrame` with size-threshold bypass

⚡ **Performance**
- Tail-window reparsing: only re-lex the trailing unstable suffix on append-only streams
- Stable token boundary detection (headings, closed fences, double newlines)
- Reference-link safety guard to prevent incorrect tail-window optimization

📚 **Documentation**
- Chunked/concat mode toggle on streaming docs page with per-SDK examples
- Stream mode toggle and virtual list tip on LlmStreaming demo component
- Offset chunk semantics, batched flushing docs, and API reference updates
- Examples CTA added to homepage hero
- LLM streaming mode added to competitor comparison data

🔧 **Refactoring**
- Extract `teardownStreamingBuffers()` to consolidate streaming state cleanup
- Use named `MotionA` import for tree-shaking in docs homepage
- Remove dead `IncrementalParser.reset()` method and unreachable flush call
- Fix `tokens` derivation for `streaming=true` + async extensions case

🧪 **Testing**
- 20+ unit tests for imperative streaming API (append, offset, mode switching, warnings, edge cases)
- Expanded imperative streaming E2E test page with metrics, controls, and offset panel
- Tail-window reparsing tests including reference-link safety
- Stream benchmark utility for measuring per-chunk parse performance

📦 **Chore**
- Bump svelte, vite, vitest, typescript-eslint, katex, and other dependencies
- Fix pre-commit hook treating `trunk fmt` exit code 1 (files formatted) as failure

## Commits

- [`272bbc1`](https://github.com/humanspeak/svelte-markdown/commit/272bbc17fc9ef5b0473ee7bcfeffdfa3bf300ad8) refactor: extract teardownStreamingBuffers and remove dead code
- [`05542ab`](https://github.com/humanspeak/svelte-markdown/commit/05542ab8f154ac5665508c7f0a3eb2fc5c777323) docs: document offset chunks, batched flushing, and tail-window parsing
- [`3df0398`](https://github.com/humanspeak/svelte-markdown/commit/3df03981e982235d052ac81230190300bf41b5d8) test: expand imperative streaming test page with metrics and controls
- [`db5bfaa`](https://github.com/humanspeak/svelte-markdown/commit/db5bfaaf72c2a3443e998f9cb13659bad75ec333) perf: add tail-window reparsing and batched append flushing
- [`4f25089`](https://github.com/humanspeak/svelte-markdown/commit/4f25089b97b23828d38ec09e62a1b56de71e47b4) refactor(docs): use named MotionA import for tree-shaking
- [`de7b961`](https://github.com/humanspeak/svelte-markdown/commit/de7b961357141feb4c105f5a18c2e371ea631c5f) docs: add LLM streaming mode to competitor comparison data
- [`8f421b9`](https://github.com/humanspeak/svelte-markdown/commit/8f421b928bb7c4c35dad013ec7e5b7818abc8e7d) docs: add Examples CTA and use imperative streaming on homepage demo
- [`17677e3`](https://github.com/humanspeak/svelte-markdown/commit/17677e3246ae0101e3e4730688b214ec9c575c9b) chore: bump dependencies and fix pre-commit hook exit code handling
- [`99d8dac`](https://github.com/humanspeak/svelte-markdown/commit/99d8dac224e08ffc4b2809114b3aaf415b31daaa) docs: add chunked/concat toggle to streaming docs and demo
- [`1f9c639`](https://github.com/humanspeak/svelte-markdown/commit/1f9c63905d1bb4b017ee4dda9217cb85ae9cae4b) feat: add imperative streaming API with writeChunk and resetStream
- [`f2a0737`](https://github.com/humanspeak/svelte-markdown/commit/f2a07373d8353e858e9b765f00b52fa1638d3084) chore: add typescript to .ncuskip to prevent major version bumps

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: docs/src/routes/+page.svelte [hunk 1]
```diff
@@ -21,6 +21,11 @@
     import { competitors } from '$lib/compare-data'
     import type { IconName } from '$lib/icons'
 
+    interface StreamingMarkdownHandle {
+        writeChunk: (chunk: string) => void
+        resetStream: (nextSource?: string) => void
+    }
```
[the remark this answers is not in the document]
Comment (jaysin586) [reply docs/src/routes/+page.svelte:27]: Good callout! For a docs homepage page though, we intentionally keep the interface inline to avoid adding imports that would clutter the example and make it harder to scan. The actual SDK examples in the streaming docs page do use the exported types.

File: src/lib/utils/incremental-parser.test.ts [hunk 2]
```diff
@@ -153,4 +139,83 @@ describe('IncrementalParser', () => {
...
+            const lexSpy = vi.spyOn(parseAndCacheModule, 'lexAndClean')
+            const parser = new IncrementalParser(createDefaultOptions())
+            const source = '[foo]\n\nTail'
+            const appended = `${source}\n\n[foo]: /docs`
+
+            parser.update(source)
+            parser.update(appended)
+
+            expect(lexSpy.mock.calls[1]?.[0]).toBe(appended)
+        })
```
[the remark this answers is not in the document]
Comment (jaysin586) [reply src/lib/utils/incremental-parser.test.ts:205]: Great catch — fixed in f968a49. The test now captures `result` from the second `parser.update()` and asserts `result.divergeAt === 0`. We also went further and made the runtime force `divergeAt=0` whenever reference syntax is detected in either the previous or current source, so the rerender is guaranteed.

File: src/lib/utils/incremental-parser.ts [hunk 3]
```diff
@@ -13,6 +13,16 @@ import type { SvelteMarkdownOptions } from '$lib/types.js'
...
 import { lexAndClean } from '$lib/utils/parse-and-cache.js'
 
+interface TailWindowBoundary {
+    prefixCount: number
+    reparseOffset: number
+}
+
+const CLOSED_FENCE_RE = /^ {0,3}(`{3,}|~{3,}).*\n[\s\S]*\n {0,3}\1[ \t]*\n*$/
+const LINK_REFERENCE_RE = /\[[^\]\n]+\]\[[^\]\n]*\]/
+const SHORTCUT_REFERENCE_RE = /\[[^\]\n]+\](?![\[(])/ // Excludes inline links/images and full refs
```
[the remark this answers is not in the document]
Comment (jaysin586) [reply src/lib/utils/incremental-parser.ts]: Already fixed in a276bf1 — removed the unnecessary `\\[` escape inside the character class. The current regex is `/\\[[^\\]\\n]+\\](?![[(])/`.

File: src/lib/utils/incremental-parser.ts [hunk 4]
```diff
@@ -48,16 +58,95 @@ export class IncrementalParser {
...
     /**
      * Creates a new incremental parser instance.
      *
      * @param options - Svelte markdown parser options forwarded to Marked's Lexer
      */
     constructor(options: SvelteMarkdownOptions) {
         this.options = options
+        this.tailWindowDisabled =
+            typeof options.walkTokens === 'function' || options.tokenizer != null
+    }
```
[the remark this answers is not in the document]
Comment (jaysin586) [reply src/lib/utils/incremental-parser.ts:89]: Fixed in f968a49. The `tailWindowDisabled` constructor now also checks `options.extensions.block` and `options.extensions.inline` for registered tokenizer hooks. When Marked processes extensions via `new Marked(...extensions).defaults`, block/inline tokenizers land in `options.extensions.block[]` and `options.extensions.inline[]` respectively, so this covers mermaid, footnotes, and other custom block-level extensions.

File: src/lib/utils/stream-benchmark.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    return {
+        totalChars: source.length,
+        chunkCount: chunks.length,
+        totalParseMs,
+        peakParseMs: parseDurationsMs.length > 0 ? Math.max(...parseDurationsMs) : 0,
+        p95ParseMs: percentile(parseDurationsMs, 0.95),
+        finalTokens,
+        parseDurationsMs
+    }
+}
```
[the remark this answers is not in the document]
Comment (jaysin586) [reply src/lib/utils/stream-benchmark.ts:76]: Already addressed in c6b90b8 — both `percentile` and `benchmarkAppendStream` now have full JSDoc with `@param`, `@returns`, and `@example` tags.

File: docs/src/routes/docs/advanced/llm-streaming/+page.svx [hunk 6]
```diff
@@ -98,8 +252,49 @@ The component works reactively by default. For best streaming performance, pass
...
+
+        try {
+            while (true) {
+                const { done, value } = await reader.read()
+                if (done) break
+                markdown?.writeChunk(decoder.decode(value, { stream: true }))
+            }
+        } finally {
+            reader.releaseLock()
+        }
```
[the remark this answers is not in the document]
Comment (jaysin586) [reply docs/src/routes/docs/advanced/llm-streaming/+page.svx:289]: Fair point — the section title is a bit misleading. This example intentionally demonstrates raw chunked streaming via `ReadableStream`, which is the most common pattern for LLM API endpoints (including OpenAI and Anthropic's streaming HTTP responses). These aren't true SSE with `text/event-stream` framing — they're `Transfer-Encoding: chunked` responses. We'll rename the section to "With fetch and Chunked Responses" in a follow-up to avoid confusion.

**Comment by jaysin586** [comment]
## Responses to CodeRabbit outside-diff comments

**1. Force rerender when reference definitions change existing tokens** (`incremental-parser.ts:159-173`)
Fixed in f968a49. The `update()` method now checks both `prevSource` and the new `source` for reference syntax via `hasAppendSensitiveReferenceSyntax()`. When detected, `divergeAt` is forced to `0`, ensuring all tokens are rerendered even when `token.raw` hasn't changed (since reference definitions alter inline children).

**2. Keep `reset()` as a deprecated compatibility shim** (`incremental-parser.ts:57-79`)
Not addressing. `IncrementalParser` is an internal utility — while it's exported from `index.ts`, `reset()` was never part of the documented public API and has no external consumers. The component manages parser lifecycle by creating new instances via `applyStreamingSource(nextSource, true)`. Adding a deprecated shim would add dead code with no practical benefit.

**3. Add streaming types to type exports summary** (`types/+page.svx:439-447`)
Fixed in f968a49. `StreamingChunk` and `StreamingOffsetChunk` are now included in the type exports code block in the docs.
